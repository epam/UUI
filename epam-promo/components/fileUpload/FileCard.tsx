import * as React from 'react';
import css from './FileCard.scss';
import { i18n } from '../..';
import { cx, FileUploadResponse, formatBytes, IClickable, IHasCX, uuiMod } from '@epam/uui-core';
import { SvgCircleProgress } from './';
import { FlexCell, FlexRow, IconButton, IconContainer, Text, Tooltip } from '../';
import { ReactComponent as RemoveIcon } from '@epam/assets/icons/common/navigation-close-18.svg';
import { ReactComponent as DocIcon } from '../../icons/fileUpload/file-file_word-24.svg';
import { ReactComponent as ExelIcon } from '../../icons/fileUpload/file-file_excel-24.svg';
import { ReactComponent as PdfIcon } from '../../icons/fileUpload/file-file_pdf-24.svg';
import { ReactComponent as ImgIcon } from '../../icons/fileUpload/file-file_image-24.svg';
import { ReactComponent as VideoIcon } from '../../icons/fileUpload/file-file_video-24.svg';
import { ReactComponent as TableIcon } from '../../icons/fileUpload/file-file_table-24.svg';
import { ReactComponent as TextIcon } from '../../icons/fileUpload/file-file_text-24.svg';
import { ReactComponent as MailIcon } from '../../icons/fileUpload/file-file_eml-24.svg';
import { ReactComponent as FileIcon } from '../../icons/fileUpload/file-file-24.svg';
import { ReactComponent as ErrorIcon } from '../../icons/notification-error-fill-10.svg';

export interface FileCardItem extends Partial<File & FileUploadResponse> {
    progress?: number;
    abortXHR?: () => void;
}

export interface FileCardProps extends IClickable, IHasCX {
    file: FileCardItem;
    width?: number;
}

export const FileCard = React.forwardRef<HTMLDivElement, FileCardProps>((props, ref) => {
    const [isLoadingShow, setIsLoadingShow] = React.useState<boolean>(true);

    const { file } = props;

    const getIcon = (extension: string) => {
        switch (extension) {
            case 'doc':
            case 'docx':
                return <IconContainer size={24} icon={DocIcon} cx={css.docColor} />;
            case 'xls':
            case 'xlsx':
                return <IconContainer size={24} icon={ExelIcon} cx={css.xlsColor} />;
            case 'pdf':
                return <IconContainer size={24} icon={PdfIcon} cx={css.pdfColor} />;
            case 'gif':
            case 'jpg':
            case 'jpeg':
            case 'svg':
            case 'png':
            case 'webp':
                return <IconContainer size={24} icon={ImgIcon} cx={css.imgColor} />;
            case 'avi':
            case 'mov':
            case 'mp4':
            case 'wmw':
            case 'mkv':
                return <IconContainer size={24} icon={VideoIcon} cx={css.movieColor} />;
            case 'csv':
            case 'xml':
                return <IconContainer size={24} icon={TableIcon} cx={css.xmlColor} />;
            case 'rtf':
            case 'txt':
                return <IconContainer size={24} icon={TextIcon} cx={css.textColor} />;
            case 'eml':
            case 'emlx':
                return <IconContainer size={24} icon={MailIcon} cx={css.emlColor} />;
            default:
                return <IconContainer size={24} icon={FileIcon} cx={css.defaultColor} />;
        }
    };

    const {
        cx: componentCx,
        width,
        file: { progress, size, name, extension, error, abortXHR },
        onClick,
    } = props;
    const fileExtension = extension || name?.split('.').pop();
    const fileName = name?.split('.').slice(0, -1).join('');
    const isLoading = progress < 100;
    const isCrossShow = ((!isLoadingShow && isLoading) || !isLoading) && onClick;

    const mouseLeaveHandler = (e: React.MouseEvent) => {
        e.stopPropagation();
        setIsLoadingShow(() => true);
    };

    const mouseEnterHandler = (e: React.MouseEvent) => {
        e.stopPropagation();
        setIsLoadingShow(() => false);
    };

    const renderErrorContent = () => (
        <Tooltip trigger="hover" content={file.error.message} placement="bottom-start">
            <div className={css.errorBlock}>
                <ErrorIcon />
                {'Upload failed'}
            </div>
        </Tooltip>
    );

    const renderSuccessfulContent = () => (
        <Text size="18" fontSize="14" lineHeight="18" color="gray60">
            {fileExtension && `${fileExtension.toUpperCase()}, `}
            {isLoading && formatBytes((size / 100) * progress) + i18n.fileCard.fileSizeProgress}
            {formatBytes(size)}
        </Text>
    );

    const removeHandler = () => {
        progress && progress < 100 && abortXHR();
        onClick();
    };

    return (
        <FlexCell
            ref={ref}
            cx={cx(css.fileCardWrapper, isLoading && uuiMod.loading, componentCx, error?.isError && css.errorCardWrapper)}
            minWidth={width}
            width={!width ? '100%' : undefined}
        >
            <FlexRow cx={css.fileCardRow} size="36" alignItems="top" spacing="6">
                {fileExtension && getIcon(fileExtension)}
                <FlexCell width="100%">
                    <Text size="18" fontSize="14" lineHeight="18" color={progress < 100 ? 'gray60' : 'gray80'} cx={css.fileName}>
                        {fileName}
                    </Text>
                    {error?.isError ? renderErrorContent() : renderSuccessfulContent()}
                </FlexCell>
                <div className={cx(css.iconsBlock)} onMouseEnter={mouseEnterHandler} onMouseLeave={mouseLeaveHandler}>
                    {isLoadingShow && isLoading && <SvgCircleProgress progress={progress} size={18} />}
                    {isCrossShow && <IconButton icon={RemoveIcon} onClick={removeHandler} />}
                </div>
            </FlexRow>
        </FlexCell>
    );
});
