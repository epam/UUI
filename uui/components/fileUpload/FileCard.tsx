import * as React from 'react';
import css from './FileCard.module.scss';
import { i18n } from '../../i18n';
import { cx, FileUploadResponse, formatBytes, IClickable, IHasCX, uuiMod } from '@epam/uui-core';
import { SvgCircleProgress } from './SvgCircleProgress';
import { IconContainer } from '@epam/uui-components';
import { FlexCell, FlexRow } from '../layout';
import { IconButton } from '../buttons';
import { Text } from '../typography';
import { Tooltip } from '../overlays';
import { fileIcons } from '../../icons/icons';
import { ReactComponent as RemoveIcon } from '@epam/assets/icons/navigation-close-outline.svg';
import { ReactComponent as ErrorIcon } from '@epam/assets/icons/notification-error-fill.svg';

export interface FileCardItem extends Partial<File & FileUploadResponse> {
    /**
    * Defines upload progress. It should be numbered value in range from 0 to 100', where 100 means that file fully loaded.
    */
    progress?: number;
    /**
    * Callback to handle aborted event which fired when a request has been aborted.
    * See {@link https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/abort_event} for details
    */
    abortXHR?: () => void;
}

export interface FileCardProps extends IClickable, IHasCX {
    /** Defines file card item */
    file: FileCardItem;
    /**
     * Defines card width
     * @default '100%'
     */
    width?: number;
}

export const FileCard = /* @__PURE__ */React.forwardRef<HTMLDivElement, FileCardProps>((props, ref) => {
    const [isLoadingShow, setIsLoadingShow] = React.useState<boolean>(true);

    const { file } = props;

    const getIcon = (extension: string) => {
        switch (extension) {
            case 'doc':
            case 'docx':
                return <IconContainer size={ 24 } icon={ fileIcons.docIcon } cx={ css.docColor } />;
            case 'xls':
            case 'xlsx':
                return <IconContainer size={ 24 } icon={ fileIcons.exelIcon } cx={ css.xlsColor } />;
            case 'pdf':
                return <IconContainer size={ 24 } icon={ fileIcons.pdfIcon } cx={ css.pdfColor } />;
            case 'gif':
            case 'jpg':
            case 'jpeg':
            case 'svg':
            case 'png':
            case 'webp':
                return <IconContainer size={ 24 } icon={ fileIcons.imgIcon } cx={ css.imgColor } />;
            case 'avi':
            case 'mov':
            case 'mp4':
            case 'wmw':
            case 'mkv':
                return <IconContainer size={ 24 } icon={ fileIcons.videoIcon } cx={ css.movColor } />;
            case 'csv':
            case 'xml':
                return <IconContainer size={ 24 } icon={ fileIcons.tableIcon } cx={ css.defaultColor } />;
            case 'rtf':
            case 'txt':
                return <IconContainer size={ 24 } icon={ fileIcons.textIcon } cx={ css.defaultColor } />;
            case 'eml':
            case 'emlx':
                return <IconContainer size={ 24 } icon={ fileIcons.mailIcon } cx={ css.defaultColor } />;
            default:
                return <IconContainer size={ 24 } icon={ fileIcons.fileIcon } cx={ css.defaultColor } />;
        }
    };

    const {
        cx: componentCx,
        width,
        file: {
            progress, size, name, extension, error, abortXHR,
        },
        onClick,
    } = props;
    const fileExtension = extension || name?.split('.').pop();
    const fileName = name?.split('.').slice(0, -1).join('');
    const isLoading = progress < 100;
    const isCrossShow = ((!isLoadingShow && isLoading) || !isLoading) && onClick;

    const mouseLeaveHandler = (e: React.MouseEvent) => {
        e.stopPropagation();
        setIsLoadingShow(() => true);
    };

    const mouseEnterHandler = (e: React.MouseEvent) => {
        e.stopPropagation();
        setIsLoadingShow(() => false);
    };

    const renderErrorContent = () => (
        <Tooltip content={ file.error.message } placement="bottom-start">
            <div className={ css.errorBlock }>
                <IconContainer icon={ ErrorIcon } size={ 12 } cx={ css.defaultColor } />
                {i18n.fileCard.failedUploadErrorMessage}
            </div>
        </Tooltip>
    );

    const renderSuccessfulContent = () => (
        <Text size="18" fontSize="14" lineHeight="18" color="tertiary">
            {fileExtension && `${fileExtension.toUpperCase()}, `}
            {isLoading && formatBytes((size / 100) * progress, 2, false) + i18n.fileCard.fileSizeProgress}
            {formatBytes(size)}
        </Text>
    );

    const removeHandler = () => {
        progress && progress < 100 && abortXHR();
        onClick();
    };

    return (
        <FlexCell
            ref={ ref }
            cx={ cx(css.root, 'uui-file_card', (isLoading || error?.isError) && uuiMod.loading, componentCx, error?.isError && uuiMod.error) }
            minWidth={ width }
            width={ !width ? '100%' : undefined }
        >
            <FlexRow size="36" alignItems="top" columnGap="6">
                {fileExtension && getIcon(fileExtension)}
                <FlexCell width="100%">
                    <Text size="18" fontSize="14" lineHeight="18" color={ (progress < 100 || error?.isError) ? 'tertiary' : 'primary' } cx={ css.fileName }>
                        {fileName}
                    </Text>
                    {error?.isError ? renderErrorContent() : renderSuccessfulContent()}
                </FlexCell>
                <div className={ cx(css.iconsBlock) } onMouseEnter={ mouseEnterHandler } onMouseLeave={ mouseLeaveHandler }>
                    {isLoadingShow && isLoading && <SvgCircleProgress progress={ progress } size={ 18 } />}
                    {isCrossShow && <IconButton icon={ RemoveIcon } onClick={ removeHandler } />}
                </div>
            </FlexRow>
        </FlexCell>
    );
});
