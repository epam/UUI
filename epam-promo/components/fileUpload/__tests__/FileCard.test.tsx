import React from 'react';
import { FileCard } from '../FileCard';
import renderer from 'react-test-renderer';

describe('FileCard', () => {
    it('should be rendered correctly', () => {
        const tree = renderer
            .create(
                <FileCard
                    file={{
                        id: 1,
                        name: 'Test.xls',
                        size: 12546,
                        progress: 0,
                    }}
                />
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should be rendered correctly', () => {
        const tree = renderer
            .create(
                <FileCard
                    file={{
                        id: 1,
                        name: 'Test.doc',
                        size: 12546,
                        progress: 0,
                    }}
                    width={140}
                    onClick={jest.fn}
                />
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should be rendered correctly', () => {
        const tree = renderer
            .create(
                <FileCard
                    file={{
                        id: 1,
                        name: 'Test.gif',
                        size: 12546,
                        progress: 100,
                    }}
                    width={140}
                    onClick={jest.fn}
                />
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});
