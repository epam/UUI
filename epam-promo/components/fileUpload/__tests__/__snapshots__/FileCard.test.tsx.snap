// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileCard should be rendered correctly 1`] = `
<div
  className="fileCardWrapper uui-loading"
  style={
    Object {
      "alignSelf": undefined,
      "flexBasis": "100%",
      "flexGrow": undefined,
      "flexShrink": undefined,
      "minWidth": 0,
      "textAlign": undefined,
    }
  }
>
  <div
    className="root size-36 background-none spacing-6 fileCardRow container align-items-top"
  >
    <div
      className="container uui-icon uui-enabled root icon-container-color-gray60 xlsColor"
      style={Object {}}
    >
      <svg
        className=""
        height={24}
        width={24}
      />
    </div>
    <div
      className=""
      style={
        Object {
          "alignSelf": undefined,
          "flexBasis": "100%",
          "flexGrow": undefined,
          "flexShrink": undefined,
          "minWidth": 0,
          "textAlign": undefined,
        }
      }
    >
      <div
        className="root font-sans text-color-gray60 line-height-18 font-size-14 v-padding-0 fileName container"
      >
        Test
      </div>
      <div
        className="root font-sans text-color-gray60 line-height-18 font-size-14 v-padding-0 container"
      >
        XLS, 
        0 Bytes of 
        12.25 KB
      </div>
    </div>
    <div
      className=""
      style={
        Object {
          "alignSelf": undefined,
          "flexBasis": "0",
          "flexGrow": undefined,
          "flexShrink": undefined,
          "minWidth": "18px",
          "textAlign": undefined,
        }
      }
    >
      <svg
        className="root"
        height={18}
        width={18}
      >
        <circle
          cx={9}
          cy={9}
          fill="transparent"
          r={3}
          stroke="#ACAFBF"
          strokeDasharray={18.84955592153876}
          strokeDashoffset={18.84955592153876}
          strokeWidth={6}
        />
        <circle
          cx={9}
          cy={9}
          fill="transparent"
          r={8}
          stroke="#ACAFBF"
          strokeWidth={1}
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`FileCard should be rendered correctly 2`] = `
<div
  className="fileCardWrapper uui-loading"
  style={
    Object {
      "alignSelf": undefined,
      "flexBasis": "0",
      "flexGrow": undefined,
      "flexShrink": undefined,
      "minWidth": "140px",
      "textAlign": undefined,
    }
  }
>
  <div
    className="root size-36 background-none spacing-6 fileCardRow container align-items-top"
  >
    <div
      className="container uui-icon uui-enabled root icon-container-color-gray60 docColor"
      style={Object {}}
    >
      <svg
        className=""
        height={24}
        width={24}
      />
    </div>
    <div
      className=""
      style={
        Object {
          "alignSelf": undefined,
          "flexBasis": "100%",
          "flexGrow": undefined,
          "flexShrink": undefined,
          "minWidth": 0,
          "textAlign": undefined,
        }
      }
    >
      <div
        className="root font-sans text-color-gray60 line-height-18 font-size-14 v-padding-0 fileName container"
      >
        Test
      </div>
      <div
        className="root font-sans text-color-gray60 line-height-18 font-size-14 v-padding-0 container"
      >
        DOC, 
        0 Bytes of 
        12.25 KB
      </div>
    </div>
    <div
      className=""
      style={
        Object {
          "alignSelf": undefined,
          "flexBasis": "0",
          "flexGrow": undefined,
          "flexShrink": undefined,
          "minWidth": "18px",
          "textAlign": undefined,
        }
      }
    >
      <svg
        className="root"
        height={18}
        width={18}
      >
        <circle
          cx={9}
          cy={9}
          fill="transparent"
          r={3}
          stroke="#ACAFBF"
          strokeDasharray={18.84955592153876}
          strokeDashoffset={18.84955592153876}
          strokeWidth={6}
        />
        <circle
          cx={9}
          cy={9}
          fill="transparent"
          r={8}
          stroke="#ACAFBF"
          strokeWidth={1}
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`FileCard should be rendered correctly 3`] = `
<div
  className="fileCardWrapper uui-loading"
  style={
    Object {
      "alignSelf": undefined,
      "flexBasis": "0",
      "flexGrow": undefined,
      "flexShrink": undefined,
      "minWidth": "140px",
      "textAlign": undefined,
    }
  }
>
  <div
    className="root size-36 background-none spacing-6 fileCardRow container align-items-top"
  >
    <div
      className="container uui-icon uui-enabled root icon-container-color-gray60 imgColor"
      style={Object {}}
    >
      <svg
        className=""
        height={24}
        width={24}
      />
    </div>
    <div
      className=""
      style={
        Object {
          "alignSelf": undefined,
          "flexBasis": "100%",
          "flexGrow": undefined,
          "flexShrink": undefined,
          "minWidth": 0,
          "textAlign": undefined,
        }
      }
    >
      <div
        className="root font-sans text-color-gray80 line-height-18 font-size-14 v-padding-0 fileName container"
      >
        Test
      </div>
      <div
        className="root font-sans text-color-gray60 line-height-18 font-size-14 v-padding-0 container"
      >
        GIF, 
        12.25 KB of 
        12.25 KB
      </div>
    </div>
    <div
      className=""
      style={
        Object {
          "alignSelf": undefined,
          "flexBasis": "0",
          "flexGrow": undefined,
          "flexShrink": undefined,
          "minWidth": "18px",
          "textAlign": undefined,
        }
      }
    >
      <svg
        className="root"
        height={18}
        width={18}
      >
        <circle
          cx={9}
          cy={9}
          fill="transparent"
          r={3}
          stroke="#ACAFBF"
          strokeDasharray={18.84955592153876}
          strokeDashoffset={0}
          strokeWidth={6}
        />
        <circle
          cx={9}
          cy={9}
          fill="transparent"
          r={8}
          stroke="#ACAFBF"
          strokeWidth={1}
        />
      </svg>
    </div>
  </div>
</div>
`;
